python/ray/tune/tests/test_multinode_sync.py

```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
+                        for f in os.listdir(self._indicator_dir)
+                        if f.startswith("indicator")
+                    ]
+                    if len(indicators) >= self._num_indicators:
+                        raise RuntimeError("All the indicators are there.")
+
+        with self.assertRaises(RuntimeError):
+            tune.run(
+                train,
+                name="checkpoint_test",
+                num_samples=3,
+                resources_per_trial={"cpu": 4},
+                max_failures=0,
+                local_dir="/cluster/node",
+                trial_name_creator=lambda trial: trial.trial_id,
+                trial_dirname_creator=lambda trial: trial.trial_id,
+                keep_checkpoints_num=2,
+                callbacks=[FailOnIndicator("/cluster/shared", num_indicators=3)],
+                verbose=2,
+            )
+
+        nodes_dir = os.path.join(self.cluster.cluster_dir, "nodes")
+
+        def get_trial_path(ip_to_trial_dir, node) -> str:
+            node_id = node["NodeID"]
+            node_ip = node["NodeManagerAddress"]
+            node_trial_id = ip_to_trial_dir[node_ip]
+
+            node_dir = os.path.join(nodes_dir, node_id)
+
+            node_trial_dir = os.path.join(node_dir, "checkpoint_test", node_trial_id)
+            return node_trial_dir
+
+        def get_checkpoint_dirs(node_trial_path: str) -> List[str]:
+            return [
+                path
+                for path in os.listdir(node_trial_path)
+                if path.startswith("checkpoint_")
+            ]
+
+        analysis = tune.ExperimentAnalysis(
+            os.path.join(nodes_dir, FAKE_HEAD_NODE_ID, "checkpoint_test")
+        )
+        ip_to_trial_dir = {
+            trial.last_result["node_ip"]: trial.trial_id for trial in analysis.trials
+        }
+
+        for node in ray.nodes():
+            node_trial_dir = get_trial_path(ip_to_trial_dir, node)
+            checkpoint_dirs = get_checkpoint_dirs(node_trial_dir)
+
+            self.assertSequenceEqual(
+                sorted(checkpoint_dirs), ["checkpoint_000008", "checkpoint_000009"]
+            )
+
+        analysis = tune.run(
+            train,
+            name="checkpoint_test",
+            resources_per_trial={"cpu": 4},
+            local_dir="/cluster/node",
+            keep_checkpoints_num=2,
+            resume="AUTO",
+            verbose=2,
+        )
+
+        ip_to_trial_dir = {
+            trial.last_result["node_ip"]: trial.trial_id for trial in analysis.trials
+        }
+
+        for node in ray.nodes():
+            node_trial_dir = get_trial_path(ip_to_trial_dir, node)
+            checkpoint_dirs = get_checkpoint_dirs(node_trial_dir)
+
+            self.assertNotIn("checkpoint_000017", checkpoint_dirs)
+            self.assertIn("checkpoint_000018", checkpoint_dirs)
+            self.assertIn("checkpoint_000019", checkpoint_dirs)
+
+            self.assertLessEqual(len(checkpoint_dirs), 4)
```

Comment:
why 4? Are you saying this is the case if after resume the trial happens to be placed onto the same physical node again and we will have _00008, _00009, _00018 and _00019?
If not this case, we will have just the 18 and 19?
It would be good to add some comments. It's kind of hard to follow.

Reply:
Added a comment